src/main/java/com/google/devtools/build/lib/actions/cache/CompactPersistentActionCache.java

```diff
@@ -512,10 +539,51 @@ public synchronized void dump(PrintStream out) {
 
       byte[] usedClientEnvDigest = MetadataDigestUtils.read(source);
 
+      int numOutputFiles = VarInt.getVarInt(source);
+      Map<String, FileArtifactValue> outputFiles = new HashMap<>(numOutputFiles);
+      for (int i = 0; i < numOutputFiles; i++) {
+        int id = VarInt.getVarInt(source);
+        String path = (id >= 0 ? indexer.getStringForIndex(id) : null);
+        if (path == null) {
+          throw new IOException("Corrupted file index");
+        }
+
+        byte[] bytes = new byte[VarInt.getVarInt(source)];
+        source.get(bytes);
+        FileArtifactValue metadata;
+        try {
+          metadata = (FileArtifactValue) objectCodecs.deserialize(ByteString.copyFrom(bytes));
```

Comment:
ditto here -- you can create a CodedOutputStream with both [a ByteBuffer](https://github.com/protocolbuffers/protobuf/blob/b90ec9c242b303e90811deebaa2e3e684b63de91/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L175-L177) and a [byte array slice](https://github.com/protocolbuffers/protobuf/blob/b90ec9c242b303e90811deebaa2e3e684b63de91/java/core/src/main/java/com/google/protobuf/CodedInputStream.java#L138-L141).